drivers/adc/ad7616/ad7616.c

```diff
@@ -38,30 +39,301 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *******************************************************************************/
 
+/* Datasheet: https://www.analog.com/media/en/technical-documentation/data-sheets/AD7616.pdf */
+
 /******************************************************************************/
 /***************************** Include Files **********************************/
 /******************************************************************************/
 #include <stdint.h>
-#include <stdio.h>
 #include <stdlib.h>
-#include "spi_engine.h"
-#include "axi_dmac.h"
-#include "no_os_gpio.h"
+#include <string.h>
+#include <assert.h>
+#include <stdio.h>
+
 #include "ad7616.h"
+#include "no_os_gpio.h"
 #include "no_os_error.h"
 #include "no_os_delay.h"
-#include "no_os_axi_io.h"
 #include "no_os_alloc.h"
 
+#ifdef PLATFORM_ZYNQMP
+#include "spi_engine.h"
+#include "axi_dmac.h"
+#include "no_os_axi_io.h"
+#else
+#include "no_os_spi.h"
+#endif
+
+#define GET_BIT(v, b) ((v >> b) & 1)
+#define TOGGLE_DELAY 8192
+
+/**
+ * @brief Compute the CRC for one channel. See CRC chapter in the datasheet
+ * @param data - Data to check
+ * @param crc - Initial crc
+ * @return The computed CRC
+ */
+static uint8_t crc_channel(uint16_t data, uint8_t crc)
+{
+	uint8_t crc_out;
+
+	uint8_t data_0 = GET_BIT(data, 0);
```

Comment:
no_os_test_bit from no_os_util.h is also an option here (although maybe slower)

Reply:
[Here](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:5,endLineNumber:17,positionColumn:5,positionLineNumber:17,selectionStartColumn:5,selectionStartLineNumber:17,startColumn:5,startLineNumber:17),source:'int+no_os_test_bit(int+pos,+const+volatile+void+*+addr)%0A%7B%0A%09return+(((const+int+*)addr)%5Bpos+/+32%5D+%3E%3E+pos)+%26+1UL%3B%0A%7D%0Aint+get_bit(int+pos,+const+volatile+unsigned+int+*+val)+%7B%0A++++return+(*val+%3E%3E+pos)+%26+0x1%3B%0A%7D%0A/*%0Aint+main+(void)+%7B%0A++++int+a+%3D+1+%2B+1%3B%0A++++int+b+%3D+1+%2B+1%3B%0A%0A++++a+%3D+no_os_test_bit(1,+%26a)%3B%0A++++b+%3D+get_bit(1,+%26b)%3B%0A%0A++++return+0%3B%0A%7D+*/'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:carmug1320,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-O2',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+ARM+GCC+13.2.0+(unknown-eabi)+(Editor+%231)',t:'0')),k:50,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+ARM+GCC+13.2.0+(unknown-eabi)+(Compiler+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:50,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) is a comparison. I would argue that we should keep my macro. Let me know what you to think

Reply:
it's quite a difference! let's keep your macro then :+1: